ykbh/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+use std::alloc::{alloc, dealloc, Layout};
+use std::convert::{TryFrom, TryInto};
+use ykpack::{
+    self, Constant, ConstantInt, IPlace, Local, LocalDecl, Statement, Terminator, TypeId,
+    UnsignedInt,
+};
+use yktrace::sir::SIR;
+
+const SUSIZE: usize = std::mem::size_of::<usize>();
+
+pub struct SIRInterpreter {
+    map: *mut u8,
+    layout: Layout,
+    locals: Vec<usize>,
+}
+
+impl SIRInterpreter {
+    pub fn new(local_decls: &Vec<LocalDecl>) -> Self {
+        // FIXME Soon this will be pre-computed and handed to us by SIR.
+        let mut size: usize = 0;
+        let mut locals = Vec::new();
+        for d in local_decls {
+            locals.push(size);
+            size += usize::try_from(SIR.ty(&d.ty).size()).unwrap();
+        }
+
+        // Allocate memory for the locals
+        let layout = Layout::from_size_align(usize::try_from(size).unwrap(), 8).unwrap();
+        let map = unsafe { alloc(layout) };
+
+        SIRInterpreter {
+            map,
+            layout,
+            locals,
+        }
+    }
+
+    /// Sets the current frame of the interpreter.
+    pub fn set_frame(&mut self, tio: usize) {
+        // FIXME Later this also sets other already initialised variables as well as the program
+        // counter of the interpreter.
+        let ptr = self.get_local(&Local(1)); // The trace inputs live in $1
+        unsafe {
+            std::ptr::copy(usize::to_ne_bytes(tio).as_ptr(), ptr as *mut u8, SUSIZE);
+        }
+    }
+
+    pub fn interpret(&mut self, body: &ykpack::Body) {
+        // Ignore yktrace::trace_debug.
+        if body.flags & ykpack::bodyflags::TRACE_DEBUG != 0 {
+            return;
+        }
+
+        for stmt in body.blocks[0].stmts.iter() {
+            match stmt {
+                Statement::MkRef(dest, src) => self.mkref(dest, src),
+                Statement::DynOffs { .. } => todo!("dynoffs"),
+                Statement::Store(dest, src) => self.store(dest, src),
+                Statement::BinaryOp { .. } => todo!("binop"),
+                Statement::Nop => {}
+                Statement::Unimplemented(_) | Statement::Debug(_) => todo!("Unimplemented"),
+                Statement::Cast(..) => todo!("cast"),
+                Statement::Call(..) | Statement::StorageDead(_) => unreachable!(),
+            }
+        }
+
+        match &body.blocks[0].term {
+            Terminator::Call {
+                operand: _op,
+                args: _args,
+                destination: _dest,
+            } => todo!(),
+            Terminator::Return => {}
+            t => todo!("{}", t),
+        }
+    }
+
+    /// Get the pointer to a Local.
+    fn get_local(&self, local: &Local) -> usize {
```

Comment:
I couldn't work out why the return type was `usize` until I realised that you're trying to avoid pointers. I think this method should be called `local_ptr` and return `*mut u8`.

Reply:
Fixed in 9e9ef8f.